### PR TITLE
driver-opts for network create

### PR DESCRIPTION
**Pull request by mavenugo**
As discussed https://github.com/docker/docker/pull/16910#issuecomment-147798196, am filing this PR after yanking the driver-opts from #16910.
Also renamed `labels` to `DriverOpts` in libnetwork via the vendor-in.

ping @icecrime @tiborvass

### CODE REVIEW

**Comment by thaJeztah** [comment]
I think this is intended for 1.9, so adding it to the milestone :)

**Comment by icecrime** [comment]
We're in code freeze, I let @tiborvass decide but we should only be accepting bugfixes.

**Comment by mavenugo** [comment]
@icecrime i understand. This is one of the requests that pretty much all the network plugin devs requested for. Its infact discussed in various libnetwork PRs. We couldnt get it in before because of various UX uncertainties. This also brings the network create ux inline with volume create for all the drivers. Since this is a very small patch but have a huge upside for the external plugins, I request you to consider getting this useful patch in. btw, the risk is very low because, we have been using the driver-opts backend from 1.7 release onwards for the default docker0 bridge.

**Comment by dave-tucker** [comment]
:+1: for having this in 1.9
Those affected - or at least those that I remember having asked for this are: @tomdee, @thockin, @shettyg, @squaremo, @erikh and @FlorianOtel

**Comment by shettyg** [comment]
@dave-tucker 

Labels support is very important for my plugin. Unfortunately, just the label support passed to network create does not add a lot of value for me (others may have a different requirement). I would really like labels support for both network create as well as endpoint create.

**Comment by calavera** [comment]
I think not supporting options in 1.9 is pretty much a bug, otherwise people won't be able to configure networks.

**Comment by shettyg** [comment]
On a higher level, most of us would really love to have libnetwork used for all our docker networking needs. It so happens that most of us (plugin writers) have our own networking backend with our own special needs. A little bit of extra help via labels lets us have a good integration of docker with the underlying infrastructure.

**Comment by mavenugo** [comment]
@shettyg I realized very late that the primary reason for the confusion is the word `label`. Though it carries the same map[string]string such as the `-opt`, it has a different meaning altogether.
Hence, we must be calling these driver-opts in order to differentiate itself clearly from labels.

Infact the reason this PR was split from #16910 because of that confusion.

By having this PR in, users can configure flexible driver options while creating the network. It is a first step towards later enhancements.

**Comment by shettyg** [comment]
@mavenugo 
Driver options for network create will be a welcome iniital addition.

File: api/client/network.go [hunk 1]
```diff
@@ -44,7 +45,8 @@ func (cli *DockerCli) CmdNetworkCreate(args ...string) error {
 	cmd.Var(&flIpamSubnet, []string{"-subnet"}, "Subnet in CIDR format that represents a network segment")
 	cmd.Var(&flIpamIPRange, []string{"-ip-range"}, "allocate container ip from a sub-range")
 	cmd.Var(&flIpamGateway, []string{"-gateway"}, "ipv4 or ipv6 Gateway for the master subnet")
-	cmd.Var(flIpamAux, []string{"-aux-address"}, "Auxiliary ipv4 or ipv6 addresses used by network driver")
+	cmd.Var(flIpamAux, []string{"-aux-address"}, "Auxiliary ipv4 or ipv6 addresses used by Network driver")
+	cmd.Var(flOpts, []string{"o", "-opt"}, "Set network options")
```
Comment (calavera) [inline api/client/network.go]: I know it's not part of this PR, but it really bothers me that some comment strings start with capital letters and other not. This new option could also be a little more descriptive, but I understand that options are optional :smile:

**Comment by calavera** [comment]
@mavenugo any way we can have a test case?

**Comment by mavenugo** [comment]
@calavera i added a simple test in `integration-cli/docker_api_network_test.go`.
Since these are driver specific options, the functionality is entirely driver specific and we cannot add a test case for that. All we can make sure is that the options configured by the user is properly processed at the libnetwork level and returned back correctly in the inspect (which is covered in the above test).

File: api/client/network.go (hunk 1 above)
Comment (mavenugo) [inline api/client/network.go]: sure. will fix that.

**Comment by mavenugo** [comment]
@calavera i can possibly add a dummy remote plugin and make sure the passed options infact reach the dummy remote plugin. Do you think that works ?

**Comment by vdemeester** [comment]
@mavenugo I would think so yeah :+1:

**Comment by tomdee** [comment]
A dummy remote plugin would really help with testing. During libnetwork development, remote plugins have been broken too many times because of a lack of automated testing.

And as I said in https://github.com/docker/docker/pull/16910#issuecomment-147776850 , network labels would be nice to have.

**Comment by mrjana** [comment]
@tomdee Thankfully that's not true anymore. We do have automated testing in libnetwork which employs a dummy remote plugin in many of our tests to make sure  there are no regressions in remote plugin protocol.

**Comment by tomdee** [comment]
:+1: That's good to hear!

**Comment by calavera** [comment]
@mavenugo @mrjana if we already have a dummy plugin in libnetwork, we can probably port it to use it here easily if we add it to the testutils package in libnetwork, we don't need to duplicate it :smile_cat: 

https://github.com/docker/libnetwork/tree/master/testutils

**Comment by mavenugo** [comment]
@calavera @vdemeester udpated with comments addressed.
Yes, I already had a dummy network plugin added in the existing tests and am reusing it for this -opts case.

**Comment by calavera** [comment]
awesome, thanks.

LGTM

File: vendor/src/github.com/docker/libnetwork/network.go [hunk 2]
```diff
@@ -484,19 +483,19 @@ func NetworkOptionIpam(ipamDriver string, addrSpace string, ipV4 []*IpamConf, ip
...
 			n.generic = make(map[string]interface{})
 		}
-		if labels == nil {
-			labels = make(map[string]string)
+		if opts == nil {
+			opts = make(map[string]string)
 		}
 		// Store the options
-		n.generic[netlabel.GenericData] = labels
+		n.generic[netlabel.GenericData] = opts
```
Comment (tiborvass) [inline vendor/src/github.com/docker/libnetwork/network.go]: as a TODO you should probably rename the `netlabel` package too

**Comment by tiborvass** [comment]
LGTM

**Comment by tiborvass** [comment]
I'll be cherrypicking this to 1.9 because it was originally bundled in another PR and we were the ones who asked to move it to a separate PR.

**Comment by mavenugo** [comment]
thanks @tiborvass @calavera @icecrime you just made network plugin devs very happy.

**Comment by thaJeztah** [comment]
Just asked @moxiegirl and docs-changes for this will also be in a separate PR that brings in all changes related to networking, so

docs LGTM

**Comment by vdemeester** [comment]
@mavenugo yay cool :heart_eyes_cat: :+1:

**Comment by tiborvass** [comment]
@mavenugo sorry needs a rebase :(

**Comment by calavera** [comment]
nooooooo!!!!!!!!!! :crying_cat_face:

**Comment by vdemeester** [comment]
oh noooeeesss :scream_cat:

**Comment by mavenugo** [comment]
:) done. checker PR(s), pls show some mercy.

**Comment by tiborvass** [comment]
@mavenugo conflict was only in tests?

**Comment by vdemeester** [comment]
@mavenugo @tiborvass yeah, only integration tests, my bad :scream:

**Comment by vdemeester** [comment]
It's green :tada:

**Pull request merged**